cmd/add.go

```diff
@@ -227,6 +241,38 @@ func readPaths(file string, splitter bufio.SplitFunc) []string {
 	return paths
 }
 
+// categorisePaths categorises each path in a given slice into either a
+// directory or file and appends the path to the corresponding slice.
+func categorisePaths(filesAndDirs, files, dirs []string) ([]string, []string) {
+	dirSet := make(map[string]bool)
+
+	for _, path := range filesAndDirs {
+		if pathIsDir(path) {
+			dirs = append(dirs, path)
+			dirSet[path] = true
+		}
+	}
+
+	for _, path := range filesAndDirs {
+		if !dirSet[path] && !dirSet[filepath.Dir(path)] {
```

Comment:
I think this avoids putting files that are directly inside a specified directory in to the files list, but doesn't stop files in subdirs of specified dirs.

It's probably a good idea to avoid duplication between files set and files discovered in dirs, so add a test where eg. subdir/a dir and subdir/a/subsubdir/file file are in the fofn, and the file is ignored until discovery happens and it is found by the directory walk